Add ability to cache and reuse a full Python build, not just the downloaded package

jthomale opened the pull request:
I stumbled across your travis-pyenv repo while looking for advice on how using pyenv with Travis. For my use case, I want to be able to build specific versions of CPython (like 2.7.13) in Travis rather than use the version that's supplied (e.g., 2.7.9). But CPython versions are pretty slow to build in pyenv, even if you cache the downloaded package using `$PYTHON_BUILD_CACHE_PATH`.

I played around a bit with pyenv and Travis and discovered that pyenv still works if you create a link to a full Python build that was previously built with pyenv (e.g., if you create the link in `~/.pyenv/versions`) rather than installing it from scratch with `pyenv install`. So I forked your repo and created my own branch that takes advantage of this so that pyenv doesn't have to rebuild Python versions after Travis caches them.

For PyPy it doesn't seem to make a ton of difference, as that seems to be pretty quick to build already, but for CPython it easily shaves off a couple of minutes from the overall Travis build process.

The main changes I made include:

* In setup-pyenv.sh, I removed `$PYTHON_BUILD_CACHE_PATH` and added `$PYENV_CACHE_PATH` (once you start caching the full build, you no longer need the `$PYTHON_BUILD_CACHE_PATH`, so it seemed redundant).
* After installing pyenv, it now checks the `$PYENV_CACHE_PATH` (default is still `~/.pyenv_cache`) to see if there's a cached version of whatever Python version you're installing, based on `$PYENV_VERSION`. If there isn't a cached version, it uses `pyenv install` to build and install it as usual. Then it moves that version over to the cache. At that point it creates a symlink to the cached directory in `$PYENV_ROOT/versions` and proceeds from there. If it finds the correct Python version in the cache, then it skips the `pyenv install` step and simply creates the symlink.  If it skips the install step, then it has to reinitialize pyenv after creating the symlink.
* Before creating the virtualenv, it now does a `pip install -U virtualenv` just to make sure it's installed and up-to-date. I think when I installed the latest versions of Python 3 with pyenv (since 3.4, maybe?) virtualenv wasn't included, and that screwed some things up. It may not be necessary, but it seemed like a good thing to add anyway, just in case. (Maybe it only made a difference with `pyenv-virtualenv`...)
* I made a few minor fixes where environment variables are concerned in the shell script and in the `travis.yml` file. I think in one spot you had `$PYENV_PYTHON` instead of `$PYENV_VERSION,` and in the YML file the cache directory was hardcoded to `~/.pyenv_cache` instead of using the value of the environment variable. Maybe one or two other small things.
* I tried to make sure comments were consistent with the changes I made.
* I updated the README based on all of the changes I made, and reworded some things just based on personal preference.

Note that, for my own use, in the README, I changed the URL pointing to the shell script in the sample `travis.yml` file to `https://github.com/jthomale/travis-pyenv/raw/fullbuild/setup-pyenv.sh`, in case you don't want to merge my pull request. If you do merge, then be sure to change that back to point to your release.

Anyway, I'm happy to talk about this if you have concerns or suggestions, or of course feel free to make additional edits or whatever. This is my first time making a pull request with any substantial changes to a completely external repository, so please forgive me if I've done anything incorrectly! 

And, no worries if you'd rather not merge. I tested this on a branch of one of my existing projects, and it does seem to work on actual code, but I realize this may well fall under the category of "dirty hacks." All the code I tested on was fairly vanilla CPython, and I don't know enough about the ins and outs of Python compilation to know if this would break in certain weird circumstances.

Thanks for your consideration, and thanks for making/releasing travis-pyenv, it's saved me some time and headaches.

jthomale committed:
Add caching of full Python builds

Caching an entire Python build is more efficient than only caching the
downloaded file (which is what PYTHON_BUILD_CACHE_PATH does). Once a
Python build is installed via the "pyenv install" command, this moves
the entire build to the Travis cache directory (e.g. ~/.pyenv_cache) and
links to it in ./pyenv/versions. Once it's cached, future Travis builds can
simply link to the cached version without having to re-download or
re-build it.

jthomale committed:
Bring travis.yml up-to-date with changes

This updates CACHE env
variables based on changes to setup-pyenv.sh, adds a test build for
a CPython version, and makes a few other minor modifications.

jthomale committed:
Update README based on chages to setup-pyenv.sh

jthomale committed:
Resolve issue with cached Python not activating

When running Travis tests using a previously cached Python, calls to
`pyenv global $PYENV_VERSION` were not activating the environment. I
think, since in this case the PYENV_VERSION was not actually being
installed with `pyenv install` and rather linked to an existing
directory, the necessary shim wasn't being created. Running `pyenv init`
after creating the link to the version fixed the issue.

jthomale committed:
Update README to fix URL in travis.yml example

JayH5 commented:
@jthomale hey, thanks for the PR. It looks really good.

I think there's value in this, but only for CPython, where the source is compiled.

> For PyPy it doesn't seem to make a ton of difference, as that seems to be pretty quick to build already, but for CPython it easily shaves off a couple of minutes from the overall Travis build process.

The reason for this is that pyenv downloads a pre-built PyPy, it doesn't compile from source. Trust me, compiling PyPy takes a *long* time 😅. And even built PyPy is quite big as it contains a lot of Python source code. Which means a ~20MB download of a compressed tarball which extracts to 100MB+. The tarball is what is stored in the `PYTHON_BUILD_CACHE_PATH`. So it's probably better to cache that than the extracted source.

That being said, I don't know if Travis compresses the cache anyway 😕. So maybe this is fine.

I'm trying to think of a way to get the "best of both worlds"...

JayH5 commented on .travis.yml:
```diff
@@ -4,22 +4,22 @@ env:
   - pip
   - directories:
-    - $HOME/.pyenv_cache
+    - "${PYENV_CACHE_PATH:-$HOME/.pyenv_cache}"
```
Nice

JayH5 commented on README.md:
```diff
@@ -50,4 +46,5 @@ script:
 * Installing pyenv by downloading a release tag rather than cloning the git repo can make your builds a bit faster in some cases. Set the `PYENV_RELEASE` environment variable to achieve that.
-* pyenv fails to install properly if `~/.pyenv` is present, even if the directory is empty. So if you cache any directories within `~/.pyenv` then you will probably break pyenv.
+* If you want to use `$PYENV_CACHE_PATH`, you must also set up Travis to cache this directory in your Travis configuration. Using the cache is optional, but it can greatly speed up subsequent builds.
+* pyenv fails to install properly if the `$PYENV_ROOT` is already present, even if the directory is empty. So if you set Travis to cache any directories within the pyenv root, then you will probably break pyenv. That's why Python builds are cached outside the pyenv root and then linked after pyenv is installed.
```
> That's why Python builds are...

Small nit: I would reword this slightly to `For this reason, Python builds are...`

jthomale replied:
I agree, sounds better. Thanks!

JayH5 commented on README.md:
```diff
@@ -1,14 +1,11 @@
 ## Usage
-1. Set the `$PYENV_VERSION` environment variable to the Python to install.
-2. Tell Travis to cache the `$HOME/.pyenv_cache` directory.
+1. Instead of setting a value for `python` in your `.travis.yml` build matrix, use the `$PYENV_VERSION` environment variable to set the pyenv Python version you want to use for a given build.
```
I'm not sure about prescribing this way of using the script. This suggests that the user would use `travis-pyenv` for all the build environments of the project.

While that may sometimes be the case, I think the more common use case is that someone needs to install one or two obscure Python implementations (e.g. PyPy) but is otherwise fine with using Travis' available Python versions. Hence why the example Travis does a `if [[ -n "$PYENV_VERSION" ]]; then` instead of just using the script by default.

jthomale replied:
Yes, great point! I think I didn't quite word that right. What I meant was that you can set either `python` or `$PYENV_VERSION` for any individual row/job in the build matrix, but you don't have to have both set for the same job. I.e., this:

```yml
language: python
matrix:
  include:
    - env: PYENV_VERSION='2.7.13' PYENV_VERSION_STRING='Python 2.7.13'
    - python: '3.5'
    - env: PYENV_VERSION=pypy-5.4.1 PYENV_VERSION_STRING='PyPy 5.4.1'
```
Instead of this:

```yml
language: python
matrix:
  include:
    - python: '2.7'
      env: PYENV_VERSION='2.7.13' PYENV_VERSION_STRING='Python 2.7.13'
    - python: '3.5'
    - python: 'pypy'
      env: PYENV_VERSION=pypy-5.4.1 PYENV_VERSION_STRING='PyPy 5.4.1'
```

Both versions have the same result, but the `python` values for jobs where you also have PYENV_VERSION specified I don't think are necessary (and personally I think it clutters things up a little). Without the `python` value, the system Python I think defaults to 2.7, but it doesn't actually affect the pyenv environment.

Anyway, I'll reword that. For my use case in fact I was planning to use pyenv for all of the environments, so that may have been what I was thinking when I wrote it.

jthomale commented:
@JayH5 Sure, no problem! Thank you for taking a look and for the excellent feedback.

> I think there's value in this, but only for CPython, where the source is compiled.
>
> >  For PyPy it doesn't seem to make a ton of difference, as that seems to be pretty quick to build already, but for CPython it easily shaves off a couple of minutes from the overall Travis build process.
>
> The reason for this is that pyenv downloads a pre-built PyPy, it doesn't compile from source. Trust me, compiling PyPy takes a long time 😅. 

Ah I see--that makes total sense and explains the difference in build times. As you probably figured, I have zero experience with anything other than CPython (and only just recently started using pyenv). :smiley: I agree this feature would only really be useful in cases where it does actually save some time. (I'd guess other distributions/implementations like Anaconda and Jython would be pre-compiled like PyPy?)

>And even built PyPy is quite big as it contains a lot of Python source code. Which means a ~20MB download of a compressed tarball which extracts to 100MB+. The tarball is what is stored in the PYTHON_BUILD_CACHE_PATH. So it's probably better to cache that than the extracted source.
>
> That being said, I don't know if Travis compresses the cache anyway 😕. So maybe this is fine.

Hmm, good points.

The [Travis docs](https://docs.travis-ci.com/user/caching#How-does-caching-work%3F) do say that the caching feature tars up the cache directories and uploads them to S3. It doesn't explicitly say it uses a compression option, but, looking at the [Travis log for the latest build on my branch](https://travis-ci.org/jthomale/travis-pyenv/jobs/200064206), it does look like it's compressing it as well: under the section "Setting up build cache," it shows that the cache archive file it's fetching is a .tgz file. 

Also, you didn't directly mention this, but it all made me wonder if Travis is caching *everything* from the entire build together in one file. In other words, does the size of the cache file that each job has to download grow with the number of different Python versions that you're testing against? And it appears that the answer to that is, no--each individual build job is cached in a separate file, so testing against (and caching) a bunch of different Python versions shouldn't slow down builds. Each job only stores and retrieves the Python version that it needs. (According to the [docs](https://docs.travis-ci.com/user/caching#Caches-and-build-matrices), jobs that use different environment variables get different caches, and in the logs on my builds I confirmed that each separate job is downloading a cache with a different filename, which remains the same from build number to build number--that is, the cache for 12.2 == 13.2 but 13.2 != 13.3.)

> I'm trying to think of a way to get the "best of both worlds"...

Yeah, the only combined options I can think of feel unnecessarily complex to me. Either:

1. Allow the end-user to specify  `$PYTHON_BUILD_CACHE_PATH` and/or `$PYENV_CACHE_PATH` for a given build or set of builds. If the setup script finds the first but not the second, then it does the normal pyenv install process using the `$PYTHON_BUILD_CACHE_PATH` to cache the downloaded files. If it finds the second, then it does the `$PYENV_CACHE_PATH` behavior (looks for the full build in the cache and links to it if it can).
2. Or, automatically decide whether to cache the full Python build or just the downloaded files based on the Python type specified in `$PYENV_VERSION`.

:man_shrugging: I guess, after investigating further how the Travis cache works, in practical terms I'm not necessarily seeing the downside to going ahead and always just caching the installed version of each Python, even in cases like PyPy where it doesn't really provide a performance boost over caching the archived tarball. Since Travis just creates a compressed tarball of the cache anyway, I don't think it negatively affects build times for PyPy and other pre-compiled Pythons, while at the same time providing a boost for CPython.

All of that said--your point that you created travis-pyenv in the first place so that you could install more obscure versions of Python on Travis as opposed to versions that are already built-in is well taken. Obviously targeting precise versions of CPython isn't something most people want or need to be able to do. In fact, my need only came about sort of as a result of some yak-shaving on a separate project, and maybe there's a better way to do what I was trying to do on that other project. :smile: I could just give in and switch over to using Docker and this would all be moot...

jthomale committed:
Update README with small tweaks

jerith commented:
I haven't read all the conversation, so apologies if this has been covered before.

The reason we stopped caching pyenv dirs in the first place is that occasionally a pyenv install would fail and the broken python it produced would be cached. All subsequent builds would then use the cached version instead of reinstalling and we'd have to manually clear the cache.

JayH5 commented:
@jerith: I think in this case, provided that `pyenv install` returns a non-zero exit code in the case that a build goes wrong, the script should exit before we move the build into the cached directory.

jthomale committed:
Verify build thoroughly before committing to cache

This change attempts to address a problem where a corrupt or otherwise
non-functioning Python build could be saved in the Travis CI cache. If a
Python build fails (e.g., the "pyenv install" command exits with status code
"1"), then the setup-pyenv.sh script exits with status code "1." Also, before
a cached Python is used, it gets verified. If it can't be verified, it's
cleared out of the cache and a new build is installed.

jthomale commented:
@jerith @JayH5 Agreed. In 5dba96b I added your suggested change @JayH5, and I also took a stab at adding more robust verification that a particular Python binary works (using the `PYENV_VERSION_STRING` if available or just calling a given Python binary with the --version flag). I was hoping that a given Python binary, such as one that exists in the cache, could easily be verified before moving forward so that it could detect a corrupt install and reinstall if needed. However, with CPython--at least with 3.6.0--it doesn't seem to like it when I call python directly from the cache. It complains about linked libraries missing. In which case, it fails verification and does the reinstall no matter what.

Just exiting the script if the build fails (before moving the build into the cached directory) is probably the simplest approach. I'll have to take another look tomorrow and strip out the superfluous stuff.

jthomale committed:
Improve error handling and reporting

The aim of this set of changes is to make the handling and reporting of
errors in setup-pyenv.sh as robust as possible. The script now does the
following. First it attempts to find and use a valid cached Python. If
it can't find the right version in the cache or if the Python build in
the cache is invalid, then it tries reinstalling (deleting a broken
cached existing version first). If the python build fails for any
reason, it immediately returns a status code 1. After a successful
reinstall, it tries moving the ~/.pyenv/versions/ directory to the cache
directory. If the move fails, it issues a warning but continues anyway.
It attempts to validate the newly installed version and returns a status
code 1 if that fails. As before, it does one final validation after the
virtualenv is created.

Messages are now output at each step to make it clearer what's going on
in the Travis logs. If a validation step fails, it outputs some
debugging information to help with troubleshooting.

jthomale commented:
@JayH5 @jerith See 9f0d327. After looking at it a bit more I went the opposite direction of the simpler approach. :roll_eyes: 

[edit: the eye roll is directed at myself, not you, just to be clear :grinning:]

It's a bigger set of changes, but maybe it makes the most sense in terms of managing the cache, handling errors, and helping make it easier to identify and troubleshoot problems that might arise?

FWIW it solves the problem I mentioned in the last comment by symlinking a particular (i.e. cached) Python in `~/.pyenv/versions` first and *then* attempting to verify. That logic resulted in some repetitious code, which led me to break that portion out into its own function, which eventually cascaded into the full set of changes.

JayH5 commented on setup-pyenv.sh:
```diff
@@ -10,19 +10,81 @@
+    else
+      printf "FAILED.\nClearing cached version..."
+      rm -f $version_pyenv_path
+      rm -rf $version_cache_path
```
I *think* these paths should probably be quoted.

JayH5 commented on setup-pyenv.sh:
```diff
@@ -37,21 +99,52 @@ export PATH="$PYENV_ROOT/bin:$PATH"
+    else
+      echo "** Warning: Python was succesfully built, but moving to cache"
+      echo "failed. Proceeding anyway without caching."
+    fi
```
It's not really clear what failure cases are expected here? I guess it's alright.. this is just the most complicated bit of code and I'm trying to understand what could've gone wrong at each branch.

jthomale replied:
Yeah I tried to simplify this and found it very difficult. I wasn't incredibly happy with how complex it is, either, but this was the best I could come up with that exits gracefully at each possible failure point. Overly verbose explanation of the logic:

1. First `if`: Try to use an existing cached Python. (Makes call to `use_cached_python`, which checks for and verifies a working cached version matching the requested `$PYENV_VERSION`.) If that succeeds, then we're good to go. If it fails, then ... 
2. Second `if`: Try to install the requested `$PYENV_VERSION` with pyenv. If installation fails, then report the error and exit. Otherwise, if installation succeeds, then ...
3. Third `if`: Try to move the Python that was just installed to the cache. If that `mv` command fails, for whatever reason, then issue a warning that the Python couldn't be cached, but proceed with the rest of the build anyway. Otherwise, if the `mv` command succeeds, then ...
4. Fourth `if`: Now try to use the Python that was just built and then cached. (Makes call to `use_cached_python` to verify it.) If that succeeds, then proceed. Otherwise, report the error, output debugging info, and exit with an error code.

It's possible all of those tests may not be necessary, especially the one testing to make sure the move to the cache succeeded.

JayH5 commented on setup-pyenv.sh:
```diff
@@ -10,19 +10,81 @@
+  printf "PYENV_VERSION\n$PYENV_VERSION\n"
+  printf "PYENV_VERSION_STRING\n$PYENV_VERSION_STRING\n"
+  printf "PYENV_CACHE_PATH\n$PYENV_CACHE_PATH\n"
+  echo "python --version"
```
Could do:
```
set -x
python --version
$version_cache_path/bin/python --version
which python
pyenv which python
set +x
```

JayH5 commented on setup-pyenv.sh:
```diff
@@ -10,19 +10,81 @@
+# PYENV_VERSION_STRING is set, then it validates the returned version string
+# as well (using fgrep). Returns whatever status code the command returns.
+verify_python() {
+  local python_bin="$1"
```
Personally I like to do:
```
local python_bin="$1"; shift
```
so that if the parameter isn't provided, then the function fails (at least with `set -e`).

JayH5 commented on README.md:
```diff
@@ -1,14 +1,11 @@
-2. Tell Travis to cache the `$HOME/.pyenv_cache` directory.
-3. Download and source the script in `before_install`.
+1. In your `.travis.yml` file, to run tests against non-Travis versions of Python, use the `$PYENV_VERSION` environment variable to set the pyenv Python version. (You can also set an appropriate value for `python`, as you normally would, for jobs where you want to continue running tests against Python versions that are already supported by Travis.)
+2. Tell Travis to cache the `$HOME/.pyenv_cache` directory OR (optionally) some other directory you specify in the `$PYENV_CACHE_PATH` environment variable.
```
I would prefer to keep these initial instructions as simple and short as possible so it's very straightforward to get started. Later instructions can then get into any extra details. So I'd prefer if this was closer to the original.

JayH5 reviewed (approved):
I'm not super happy about the increase in complexity but this does seem pretty thorough so I think we can merge once you've fixed one or two minor things.

JayH5 commented:
Sorry I've taken ages to review this. Dropped off my radar for a bit.

jthomale committed:
Merge changes from upstream 'develop' branch

jthomale committed:
Add 'python' version keys back to build matrix

As noted in b9221ba, the configuration change to allow Python builds to
work on Mac OS X causes Travis-CI to list jobs as Ruby unless a 'python'
key is present. My initial PR removed these from the build matrix
because I thought they were extraneous. This commit adds them back so
that Travis-CI will list the correct Python version on Linux builds.

jthomale committed:
Make stylistic corrections to setup-pyenv.sh and README.md

jthomale commented:
@JayH5 not a problem, I totally understand. I've been working on other things over the last month as well. I appreciate the thorough review. And, FWIW I completely empathize with your ambivalence about the changes given the added complexity. Thanks for approving them.

I hope you don't mind, but, in addition to making the changes you suggested, I went ahead and merged the changes from #5 with this and then added the `python` keys back into the build matrix for the non-OS X builds in `.travis.yml`. I had removed those earlier because I thought they were redundant, but with the removal of the global `language` key, it was causing Travis CI to show those as Ruby builds...

jthomale committed:
Update initial README.md instructions

Revised initial README.md instructions to more closely match original.
My main reason for changing them was to include more use cases than just
PyPy, so I turned that third sentence into a bulleted list. Hopefully this
accomplishes that goal while also keeping it simple.
